docs/user-guide/how-to-math-broadcasting.md

```diff
@@ -14,10 +14,234 @@ kernelspec:
+
++++
+
+## Missing data, heterogeneous data, and records
+
+One of the ways Awkward Arrays extend beyond NumPy is by allowing the use of `None` for missing data. These `None` values are broadcasted like empty lists:
+
+```{code-cell} ipython3
+array1 = ak.Array([[1, 2, 3], None, [4, 5]])
+array2 = ak.Array([10, 20, 30])
+
+array1 + array2
+```
+
+Another difference from NumPy is that Awkward Arrays can contain data of mixed type, such as different numbers of dimensions. If numerical values _can_ be matched across such arrays, they are:
+
+```{code-cell} ipython3
+array1 = ak.Array([[1, 2, 3], 4, 5])
+array2 = ak.Array([10, 20, 30])
+
+array1 + array2
+```
+
+Arrays containing records can also be broadcasted, though most mathematical operations cannot be applied to records. Here is an example using {func}`ak.broadcast_arrays`.
+
+```{code-cell} ipython3
+array1 = ak.Array([
+    [{"x": 1.1, "y": [1]}, {"x": 2.2, "y": [1, 2]}, {"x": 3.3, "y": [1, 2, 3]}],
+    [],
+    [{"x": 4.4, "y": [1, 2, 3, 4]}, {"x": 5.5, "y": [1, 2, 3, 4, 5]}],
+])
+array2 = ak.Array([10, 20, 30])
+
+ak.broadcast_arrays(array1, array2)
+```
+
+## Differences from NumPy broadcasting
+
+Awkward Array broadcasting is identical to NumPy broadcasting in three respects:
+
+1. arrays with the same number of dimensions must match lengths (except for length 1) exactly,
+2. length-1 dimensions expand like scalars (one to many),
+3. for arrays with different numbers of dimensions, the smaller number of dimensions is expanded to match the largest number of dimensions.
+
+Awkward Arrays with fixed-length dimensions—not "variable-length" or "ragged"—broadcast exactly like NumPy.
+
+Awkward Arrays with ragged dimensions expand the smaller number of dimensions on the left, whereas NumPy and Awkward-with-fixed-length expand the smaller number of dimensions on the right, when implementing point 3 above. This is the only difference.
+
+Here's a demonstration of NumPy broadcasting:
+
+```{code-cell} ipython3
+x = np.array([
+    [1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12],
+])
+y = np.array([
+    [[10, 20, 30, 40], [50, 60, 70, 80], [90, 100, 110, 120]],
+    [[100, 200, 300, 400], [500, 600, 700, 800], [900, 1000, 1100, 1200]],
+])
+
+x + y
+```
+
+And fixed-length Awkward Arrays made from these can be broadcasted the same way:
+
+```{code-cell} ipython3
+ak.Array(x) + ak.Array(y)
+```
+
+but only because the latter have completely regular dimensions, like their NumPy counterparts.
+
+```{code-cell} ipython3
+print(x.shape)
+print(y.shape)
+```
+
+```{code-cell} ipython3
+print(ak.Array(x).type)
+print(ak.Array(y).type)
+```
+
+In both NumPy and Awkward Array, `x` has fewer dimensions than `y`, so `x` is expanded on the _left_ from length-1 to length-2.
+
+However, if the Awkward Array has variable-length _type_, regardless of whether the actual lists have variable lengths,
+
+```{code-cell} ipython3
+print(ak.Array(x.tolist()).type)
```

Comment:
Should we prefer the `to_list` alias rather than `tolist`?

Reply:
I'm torn on this for the same reasons that we have both aliases: NumPy's name for it is `tolist` and that's quicker to type, which matters in most of the instances when you'd want to use it. So it's a fight between two conventions: consistency with NumPy names (no underscore) and consistency with a PEP standard that prefers snake case (yes underscore).

For this, for now, though, I'll leave it. I'm prioritizing "getting stuff done" over "getting stuff perfect". It's not like we're ever going to _remove_ the `tolist` alias, such that we'd regret having promoted it in the past.